dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,269 @@
+            }
+        }));
+    }
+
+    /**
+     * If we have some reason to think the currentIndex is bad, we want to move to the next host. This is done with a
+     * compareAndSet to ensure that out of order responses which signal information about a previous host don't kick
+     * us off a good one.
+     */
+    private OptionalInt markCurrentHostAsBad(int currentIndex) {
+        int nextIndex = Math.max(currentIndex + 1, 0); // we want Integer.MAX_VALUE to wrap around to zero
+
+        boolean saved = currentHost.compareAndSet(currentIndex, nextIndex);
+        return saved ? OptionalInt.of(nextIndex) : OptionalInt.empty(); // we've moved on already
+    }
+
+    interface NodeList {
+        /**
+         * Accepts positive indexes that are greater than the length of the list, returns an item modulo the length
+         * of the list.
+         */
+        LimitedChannel get(int index);
+    }
+
+    @VisibleForTesting
+    static final class ConstantNodeList implements NodeList {
+        private final List<LimitedChannel> channels;
+
+        ConstantNodeList(List<LimitedChannel> channels, Random random) {
+            /**
+             * Shuffle the initial list to ensure that clients across the fleet don't all traverse the list in the
+             * same order.  If they did, then restarting one upstream node n would shift all its traffic (from all
+             * servers) to upstream n+1. When n+1 restarts, it would all shift to n+2. This results in the disastrous
+             * situation where there might be many nodes but all clients have decided to hammer one of them.
+             */
+            this.channels = shuffleImmutableList(channels, random);
+        }
+
+        @Override
+        public LimitedChannel get(int index) {
+            return channels.get(index % channels.size());
+        }
+    }
+
+    @VisibleForTesting
+    static final class ReshufflingNodeList implements NodeList {
+        private final Ticker clock;
+        private final Random random;
+        private final long intervalWithJitter;
+        private final AtomicLong nextReshuffle;
+
+        private volatile ImmutableList<LimitedChannel> channels;
+
+        ReshufflingNodeList(List<LimitedChannel> channels, Random random, Ticker clock) {
+            this.channels = shuffleImmutableList(channels, random);
+            this.random = random;
+            this.intervalWithJitter = RESHUFFLE_EVERY
+                    .plus(Duration.ofSeconds(random.nextInt(60) - 30))
+                    .toNanos();
+            this.nextReshuffle = new AtomicLong(clock.read() + intervalWithJitter);
+            this.clock = clock;
+        }
+
+        @Override
+        public LimitedChannel get(int index) {
+            reshuffleChannelsIfNecessary();
+            return channels.get(index % channels.size());
+        }
+
+        private void reshuffleChannelsIfNecessary() {
+            if (channels.size() < 1) {
```

Comment:
If there's only one channel there's no need to shuffle either
```suggestion
            if (channels.size() <= 1) {
```

Reply:
🤦‍♂